src/dune_lang/syntax.ml

```diff
@@ -44,22 +44,79 @@ module Version = struct
 end
 
 module Supported_versions = struct
-  type t = int Int.Map.t
-
-  let to_dyn = Int.Map.to_dyn Int.to_dyn
-
-  let make = Int.Map.of_list_exn
-
-  let greatest_supported_version t = Option.value_exn (Int.Map.max_binding t)
-
-  let is_supported t (major, minor) =
+  (* The extension supported versions are declared using an explicit list of all
+     versions but stored as a map from major versions to maps from minor version
+     to dune_lang required versions *)
+  type t = Version.t Int.Map.t Int.Map.t
+
+  let to_dyn t = Int.Map.to_dyn (Int.Map.to_dyn Version.to_dyn) t
+
+  (* We convert the exposed extension version type: `(Version.t * [ `Since of
+     Version.t ]) list` which is a list of fully qualified versions paired with
+     the corresponding dune_lang version. To the internal representation:
+     `(Version.t Int.Map.t) Int.Map.t` which is a list of major versions paired
+     with lists of minor versions paires with a dune_lang version. *)
+  let make (versions : (Version.t * [ `Since of Version.t ]) list) : t =
+    let v =
+      List.fold_left versions
+        ~init:(Int.Map.empty : Version.t Int.Map.t Int.Map.t)
+        ~f:(fun major_map (v_ext, `Since v_lang) ->
+          let major, minor = v_ext in
+          let minor_map =
+            match Int.Map.find major_map major with
+            | Some m -> m
+            | None -> Int.Map.empty
+          in
+          let minor_map = Int.Map.add_exn minor_map minor v_lang in
+          Int.Map.update major_map major ~f:(function
+            | Some _minor_map -> Some minor_map
+            | None -> Some minor_map))
+    in
+    v
+
+  let remove_uncompatible_versions lang_ver =
+    Int.Map.filter_map ~f:(fun minors ->
+        let minors =
+          Int.Map.filter minors ~f:(fun min_lang -> lang_ver >= min_lang)
+        in
+        if Int.Map.is_empty minors then
+          None
+        else
+          Some minors)
+
+  let rec greatest_supported_version ?lang_ver t =
+    let open Option.O in
+    match lang_ver with
+    | Some lang_ver ->
+      let compat = remove_uncompatible_versions lang_ver t in
+      greatest_supported_version compat
```

Comment:
I was looking at map.mli in the stdlib and found `find_first` and `find_last`. it seems that we could use this function to avoid creating an intermediate map and still avoid a linear scan. WDYT?

Reply:
as you mentioned on slack, these only operate on the key so we can't use them here